jax/scipy/sparse/linalg.py

```diff
@@ -54,42 +61,123 @@ def _identity(x):
 def _cg_solve(A, b, x0=None, *, maxiter, tol=1e-5, atol=0.0, M=_identity):
 
   # tolerance handling uses the "non-legacy" behavior of scipy.sparse.linalg.cg
-  bs = _vdot_tree(b, b)
+  bs = _vdot_tree_real(b, b)
   atol2 = jnp.maximum(jnp.square(tol) * bs, jnp.square(atol))
 
   # https://en.wikipedia.org/wiki/Conjugate_gradient_method#The_preconditioned_conjugate_gradient_method
 
   def cond_fun(value):
     x, r, gamma, p, k = value
-    rs = gamma if M is _identity else _vdot_tree(r, r)
+    rs = gamma if M is _identity else _vdot_tree_real(r, r)
     return (rs > atol2) & (k < maxiter)
 
   def body_fun(value):
     x, r, gamma, p, k = value
     Ap = A(p)
-    alpha = gamma / _vdot_tree(p, Ap)
+    alpha = gamma / _vdot_tree_real(p, Ap)
     x_ = _add(x, _mul(alpha, p))
     r_ = _sub(r, _mul(alpha, Ap))
     z_ = M(r_)
-    gamma_ = _vdot_tree(r_, z_)
+    gamma_ = _vdot_tree_real(r_, z_)
     beta_ = gamma_ / gamma
     p_ = _add(z_, _mul(beta_, p))
     return x_, r_, gamma_, p_, k + 1
 
   r0 = _sub(b, A(x0))
   p0 = z0 = M(r0)
-  gamma0 = _vdot_tree(r0, z0)
+  gamma0 = _vdot_tree_real(r0, z0)
   initial_value = (x0, r0, gamma0, p0, 0)
 
   x_final, *_ = lax.while_loop(cond_fun, body_fun, initial_value)
 
   return x_final
 
 
+def _bicgstab_solve(A, b, x0=None, *, maxiter, tol=1e-5, atol=0.0, M=_identity):
+
+  # tolerance handling uses the "non-legacy" behavior of scipy.sparse.linalg.bicgstab
+  bs = _vdot_tree_real(b, b)
+  atol2 = jnp.maximum(jnp.square(tol) * bs, jnp.square(atol))
+
+  # https://en.wikipedia.org/wiki/Biconjugate_gradient_stabilized_method#Preconditioned_BiCGSTAB
+
+  def cond_fun(value):
+    x, r, *_, k = value
+    rs = _vdot_tree_real(r, r)
+    return (rs > atol2) & (k < maxiter)
+
+  def body_fun(value):
+    x, r, rhat, alpha, omega, rho, p, q, k = value
+    # rho_ = _vdot_tree(_conj(rhat), r)
+    rho_ = _vdot_tree(rhat, r)
+    beta = rho_ / rho * alpha / omega
+    p_ = _add(r, _mul(beta, _sub(p, _mul(omega, q))))
+    phat = M(p_)
+    q_ = A(phat)
+    # alpha_ = rho_ / _vdot_tree(_conj(rhat), q_)
+    alpha_ = rho_ / _vdot_tree(rhat, q_)
+    s = _sub(r, _mul(alpha_, q_))
+    # TODO(?): stop early
+    #  It requires accessing cond_fun like this, but it's
+    #  not possible with jit...
+    #  if cond_fun((x, s, r0, alpha_, omega, rho_, p_, q_, k)):
+    #    x_ = _add(x, _mul(alpha_, phat))
+    #    return x_, s, rhat, alpha_, omega, rho_, p_, q, k
+    shat = M(s)
+    t = A(shat)
+    # omega_ = _vdot_tree(_conj(s), t) / _vdot_tree_real(t, t)
+    omega_ = _vdot_tree(s, t) / _vdot_tree_real(t, t)
```

Comment:
I am concerned that this might be effectively dividing by zero if the algorithm converges in the first half of the step. I wonder if this could be related to your testing challenges?

If I'm understanding correctly, this might not be happening for `cg` only because demonator is also the termination condition on the loop.

One way to guard against this would be replace the denominator with something like `jnp.maximum(_vdot_tree_real(t, t), epsilon)`.

Reply:
I am happy to add that in as I have seen it used in other contexts, but I haven't really seen this done in scipy's fortran revcom or any other pseudocodes. That said, I have noticed that my implementation and scipy tend to have similar breakdown scenarios (usually we add breakdown checks for whether rho or omega are zero and that is returned as part of `info`). Do you recommend we add this in anyway?

Reply:
My general rule of thumb is to try not to innovate in the details of numerical methods. I'm not an expert on most of them, and it can be easy to get the details wrong.

I would definitely suggest copying the safety/breakdown checks you notice in other implementation like SciPy rather than my ideas! Let me know if you have questions about how to implement them in JAX.

Reply:
The safety/breakdown checks might require some deeper changes in jax (similar to why `info` is currently `None`). Might be a future thing to add after getting the basic `bicgstab` to work?

Reply:
You should definitely be able to add early breakdown without pervasive changes to JAX.

There are at least two ways to do it:

1. You could nest a call to `jax.lax.cond`, either returning half-way through or doing the single GMRES step.
2. You could keep track of whether early existing should happen in a boolean variable, and use `lax.select`/`jnp.where` (basically the same) to selectively update variables, e.g., compare how loopy logic for sampling from the Poisson distribution is implemented in [JAX](https://github.com/google/jax/blob/5e91965723ac16dda361435d1bfcd4dc99201537/jax/random.py#L1075-L1093) vs [NumPy](https://github.com/numpy/numpy/blob/b743bcc1c6d0ab8b26042faf5da517b69e4fbb63/numpy/random/src/distributions/distributions.c#L558-L575):

They have slightly different performance tradeoffs, but either solution should be fine here. (1) might be a little easier.

Reply:
Thanks, this is helpful!